Tests/Basic/TerminalControllerTests.swift

```diff
@@ -0,0 +1,121 @@
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import XCTest
+import Utility
+@testable import Basic
+import libc
+
+final class PseudoTerminal {
+    let master: Int32
+    let slave: Int32
+    var outStream: LocalFileOutputByteStream
+
+    init?(){
+        var master: Int32 = 0
+        var slave: Int32 = 0
+        if openpty(&master, &slave, nil, nil, nil) != 0 {
+            return nil
+        }
+        guard let outStream = try? LocalFileOutputByteStream(filePointer: fdopen(slave, "w")) else {
+            return nil
+        }
+        self.outStream = outStream
+        self.master = master
+        self.slave = slave
+    }
+
+    func readMaster(maxChars n: Int = 1000) -> String? {
+        var buf: [CChar] = [CChar](repeating: 0, count: n)
+        if read(master, &buf, n) <= 0 {
+            return nil
+        }
+        return String(cString: buf)
+    }
+
+    deinit {
+        close(master)
+        close(slave)
+    }
+}
+
+final class TerminalControllerTests: XCTestCase {
+    func testBasic() {
+        guard let pty = PseudoTerminal(), let term = TerminalController(stream: pty.outStream) else {
+            XCTFail("Couldn't create pseudo terminal.")
+            return
+        }
+
+        // Test red color.
+        term.write("hello", inColor: .red)
+        XCTAssertEqual(pty.readMaster(), "\u{1B}[31mhello\u{1B}[0m")
+        
+        // Test clearLine.
+        term.clearLine()
+        XCTAssertEqual(pty.readMaster(), "\u{1B}[2K\r")
+
+        // Test endline.
+        term.endLine()
+        XCTAssertEqual(pty.readMaster(), "\r\n")
+
+        // Test move cursor.
+        term.moveCursor(y: 3)
+        XCTAssertEqual(pty.readMaster(), "\u{1B}[3A")
+
+        // Test color wrapping.
+        var wrapped = term.wrap("hello", inColor: .noColor)
+        XCTAssertEqual(wrapped, "hello")
+
+        wrapped = term.wrap("green", inColor: .green)
+        XCTAssertEqual(wrapped, "\u{001B}[32mgreen\u{001B}[0m")
+    }
+
+    func testProgressBar() {
+        guard let pty = PseudoTerminal() else {
+            XCTFail("Couldn't create pseudo terminal.")
+            return
+        }
+
+        // Test progress bar when writing to a non tty stream.
+        let outStream = BufferedOutputByteStream()
+        var bar = createProgressBar(forStream: outStream, header: "test")
+        XCTAssertTrue(bar is SimpleProgressBar)
+
+        runProgressBar(bar)
+        XCTAssertEqual(outStream.bytes.asString, "test\n0%: 0\n1%: 1\n2%: 2\n3%: 3\n4%: 4\n5%: 5\n")
+
+        // Test progress bar when writing a tty stream.
+        bar = createProgressBar(forStream: pty.outStream, header: "TestHeader")
+        XCTAssertTrue(bar is ProgressBar)
+
+        var output = ""
+        let thread = Thread {
+            while let out = pty.readMaster() {
+                output += out
+            }
+        }
+        thread.start()
+        runProgressBar(bar)
```

Comment:
This should probably `join` on the Thread, just to make sure it is gone before test ends.